### DISCUSSION TITLE
A little language extension that would save a lot of typing

### DISCUSSION THREAD

**Discussion by dmuysers** [modula3/cm3#1143] [Ideas]
(borrowed from Oberon)
ArrayType  = ARRAY [Index {"," Index}] OF Type.
Index         = Type | ConstExpr. // 'n' same as '[0..n-1]'

**Comment by nsgnkhibdk2cls0f** [comment]
> (borrowed from Oberon) ArrayType = ARRAY [Index {"," Index}] OF Type. Index = Type | ConstExpr. // 'n' same as '[0..n-1]'

First, which Oberon? Oberon-2 or Oberon-07?

Second, please give example code, not language grammar. I don't understand language grammar (is this EBNF?).

**Reply by dmuysers** [reply]
All Oberons, and. sorry, if you don't understand, forget it... and stop giving us lessons.

**Reply by 1997tn** [reply]
Component Pascal (a dialect of Oberon-2) CP-Lang.pdf page 7:

https://github.com/pahihu/gpcp-JVM/blob/main/documentation/CP-Lang.pdf

But CM3 also supports this short hand syntax as the document said:

http://modula3.github.io/cm3/reference/complete/html/2_2_3Arrays.html

**Reply by mikanystrom-intel** [reply]
I think the point is that

ARRAY 10 OF INTEGER

is OK in Oberon

but in M3 you have to write

ARRAY [ 0 .. 9 ] OF INTEGER

for the same thing.

("10" is not a "Type")

I hate to say no, but I don't like it.  The reason is that I don't think it is really sensible to use explicit subranges like this.  They "should" almost always be constructed from constants.  I can't (and hope you don't suggest I should be able to) write

TYPE 
   Index = 10;
   T = ARRAY Index OF INTEGER;

Why do I say that they should almost always be constructed from constants?  Because wherever you *use* the index you want to declare a variable of type Index.

Consider:

TYPE
   Index = [0 .. 9];
   T = ARRAY Index OF INTEGER;
   ImportantStuff = RECORD
      index : Index;
   END;

versus

TYPE
   T = ARRAY 10 OF INTEGER;
   ImportantStuff = RECORD
      index : [FIRST(T) .. LAST(T)];
   END;

which is clearer?  I think definitely the former...

And if you are not declaring your index variables as being of type Index, then why not?  The fact that they are in fact in-range is (likely to be) an important invariant that you can easily capture using the M3 type system.  The only problem is that it doesn't work well with WHILE... but heck you can't have everything. (*)

(* I mean that in C if you have

    int arr[10];
    int *p;

p's legal range is from &(p[-1]) to &(p[10]), inclusive.  Although you are not allowed to dereference those two.  Whereas in M3, you will get a range fault if you try to set the index to FIRST(T) - 1 or LAST(T) + 1 ...

That also goes back to the Pascal tradition of running FOR loops from lo to hi inclusive, i.e., for i = 0 ; i <= n - 1.  Here maybe C has the better practice, in that one generally runs for loops with a < check, for i =0; i < n .)

**Reply by dmuysers** [reply]
Not the [0..9] case, of course. But often I have to write, e.g.:
CONST MaxIncludes = 100;
TYPE Includes = ARRAY [0..MaxIncludes-1] OF Include;
.... If n < MaxIncludes THEN ... ELSE Panic (.... etc
BTW John Gough's extensible arrays (cf. ComponentPascal for .net) would be a welcome extension

**Reply by mikanystrom-intel** [reply]
> Not the [0..9] case, of course. But often I have to write, e.g.: CONST MaxIncludes = 100; TYPE Includes = ARRAY [0..MaxIncludes-1] OF Include; .... If n < MaxIncludes THEN ... ELSE Panic (.... etc BTW John Gough's extensible arrays (cf. ComponentPascal for .net) would be a welcome extension

Ah my practice is to "un-nest" and code thus (I actually saw a YouTube about this yesterday...)

CONST 
   MaxIncludes = 100;
 
TYPE 
   Includes = ARRAY [0..MaxIncludes-1] OF Include;

...

   IF n > LAST(Includes) THEN Panic END;

   (* n is OK -- proceed *)

---

Who needs extensible arrays???  That's not the Modula-3 way.

THIS is the Modula-3 way:

INTERFACE Include;
TYPE T = ... ;
END Include.

INTERFACE IncludeSeq = Sequence(Include) END Include.

MODULE IncludeSeq = Sequence(Include) END Include.

VAR myIncludes = NEW(IncludeSeq.T).init();

... myIncludes.addhi(...) ...

**Reply by dmuysers** [reply]
The advantage of extensible arrays is that you can write:
TYPE Includes = VECTOR OF Include;
VAR InclList := NEW (Includes, 100);
APPEND (InclList, item); 
FOR i := 0 TO LENGTH (InclList) - 1 DO...
Use (InclList [i]);
END;
The accessors can be inlined, Only automatic expansion probably costs a procedure call.
And you don't have to import a generic, or worse, define a new template.
To be very honest, I am not a great fan of M3's generics. They are just glorified macros.
And, the Sequence one is just over-defined for most of its useful applications,

**Reply by dmuysers** [reply]
If I had still enough time to live, I would write a compiler for a simplified, application-oriented version of Modula
-  1:1 instead of m:n INTERFACE/MODULE relationship
-  no exceptions 
-  abstract or extensible or final OBJECTS instead of revelations.
-  builtin maps and extensible arrays 
-  pm3-like UTF-8 encoded TEXT
-  support for i18n
-  no generics
a kind of simple object-oriented 'Go'scal, the compiler would be much simpler as the cm3 one.
Even golang has become a monster after introducing generics in its recent, vastly inflated, specifications
(but its tool set is simply brillant)
Keep it as simple as possible (reminder)

**Reply by nsgnkhibdk2cls0f** [reply]
@dmuysers maybe you will like Oberon+: https://github.com/rochus-keller/Oberon

**Reply by nsgnkhibdk2cls0f** [reply]
I don't like the way generics is implemented on Modula-3 either, but I know generics is something required for a useful language. As I'm familiar with the Delphi syntax I prefer the Delphi syntax:

https://docwiki.embarcadero.com/RADStudio/Alexandria/en/Overview_of_Generics

**Reply by mikanystrom-intel** [reply]
> I don't like the way generics is implemented on Modula-3 either, but I know generics is something required for a useful language. As I'm familiar with the Delphi syntax I prefer the Delphi syntax:
> 
> https://docwiki.embarcadero.com/RADStudio/Alexandria/en/Overview_of_Generics

This is kind of a nonsensical statement along the lines of "I prefer Modula-3 interfaces over Java interfaces".  Just because they use the same word doesn't make them the same thing.

Modula-3 generics are modules, not types.  Certainly you are free to dislike them, but they aren't semantically similar to Delphi (or C++ or Java) generics in any way.

**Reply by nsgnkhibdk2cls0f** [reply]
I don't care if they are the same thing or not. On my dictionary, writing a data structure and methods but without an already known type specified that could be later re-used for other types just by plugging the type into it is called **_generics_**.

**Reply by mikanystrom-intel** [reply]
> I don't care if they are the same thing or not. On my dictionary, writing a data structure and methods but without an already known type specified that could be later re-used for other types just by plugging the type into it is called **_generics_**.

Modula-3 generics don't operate on types, neither as input nor as output.  They take modules as input (parameters) and produce modules as output.

**Reply by 1997tn** [reply]
Template class of C++, but closer to macro. The module word sounds fancy, but it's really just equivalent to class in other class based languages. On C++ and Java, each class usually lives on it own source file (just like M3's module, C++ has header (.h) and implementation (.cpp) separation, just like M3's .i3 and .m3), even though they support nested classes. Module maps well to class.

**Reply by mikanystrom-intel** [reply]
> Template class of C++, but closer to macro. The module word sounds fancy, but it's really just equivalent to class in other class based languages. On C++ and Java, each class usually lives on it own source file (just like M3's module, C++ has header (.h) and implementation (.cpp) separation, just like M3's .i3 and .m3), even though they support nested classes. Module maps well to class.

Guys, you're confusing the language *definitions* of mechanisms with *common usage models*.  They are not the same.  And this kind of wishy-washy vagueness tends to lead programming discussions completely astray.

I don't know C++ well enough to comment, but take C.  You can search high and low for "header" and "implementation" in the C language definition, and you won't find it.  It's not actually part of the language at all.  Instead, programmers have developed *conventions* about how to use mechanisms, in C in particular #include, to do things that are interesting from a software-engineering point of view.  In the case of C headers, what they are trying to do is to *simulate Modula modules*, which is a concept that otherwise does not really exist in C itself.  (One of the best books on how to write C "modules" is by the way based on Modula-3's libraries---Hanson's book.)

In the same way, Modula-3 modules are not classes, nor are Modula-3 generics classes, or types, or parameterized types.  Modula-3 modules are collections of declarations and definitions, and Modula-3 generics are actually just very cleverly written IMPORT statements (you can call them macros but in that case they are very very restricted macros, since the only thing they let you change is the "x" in IMPORT x AS y), which semantically link together these modules.

What makes it possible to use Modula-3 generics to do type templates is *conventions* among Modula-3 programmers.  In particular, the convention to define approximately one important type per interface module and call that type "T".  Nothing about the Modula-3 language definition requires anybody to code like this.

As a programmer, you are not necessarily bound by these conventions.  I remember a time when C++ programmers put most of their executable code in .h files, for example.  And you can use Modula-3 generics to do things very different from defining new object types (classes).

The main difference between modules and classes by the way is that modules, as well as their constituent components, aren't in any kind of semantically enforced type-subtype relationship to each other.  As an example, this allows you to do things with M3 generics that "look and smell" like multiple inheritance even though the M3 type system doesn't allow multiple inheritance between object types.

The most common "other thing" I have seen M3 generics used for is to define not type hierarchies but linked type hierarchies (so a generic interface module could for example define two or more derived types, which needn't be in any particular typing relationship between themselves, of two or more base types).

And yes while the M3 generic implementation *is* very simple, I don't see why this is a drawback.  Figuring out "how to name objects" is one of the most difficult problems in formal logic (of which programming languages are a subset) and having a mechanism that does that reasonably by simple means ought to be a point of pride, not something that needs to be excused.

Another comment here is that when you learn about a new programming mechanism, ask not what you can do with it---but ask what you CANNOT do with it (how does the design of this mechanism prevent me from shooting myself in the foot?)  It is very hard to shoot yourself in the foot with Modula-3's generics.  (As opposed to, say, a general unhygienic macro mechanism such as that provided by cpp.)

**Reply by nsgnkhibdk2cls0f** [reply]
M3's generics is very similar to Vector Pascal's: https://sourceforge.net/projects/vectorpascalcom/

Download it and read manual.pdf, you will see, very similar.

**Reply by mikanystrom-intel** [reply]
> M3's generics is very similar to Vector Pascal's: https://sourceforge.net/projects/vectorpascalcom/
> 
> Download it and read manual.pdf, you will see, very similar.

I am not sure why anyone on this list should care about how Vector Pascal does things.

But I am open-minded and did as you said.  And you totally missed my point (or ignored it).

Vector Pascal uses types as parameters to what it calls "parametric units".  I see nothing wrong with this.

But my point was that Modula-3 DOES NOT USE TYPES AS PARAMETERS TO GENERICS!!

MODULA-3 USES INTERFACE MODULES AS PARAMETERS TO GENERICS!

Two things are not similar just because it looks similar or has a similar name!  A photo of an elephant is not an elephant!

**Comment by dmuysers** [comment]
|  @dmuysers maybe you will like Oberon+: https://github.com/rochus-keller/Oberon
No.
1. It's GPLed, rather than MIT ot BSD
2. It did not implement the latest features of the official Active Oberon 2019

**Reply by nsgnkhibdk2cls0f** [reply]
1. What's wrong with being GPL-ed? CM3 is also GPL-ed, why are you here?
2. Rochus Keller has his own design. The language is named Oberon+, not Oberon. Why does he need to implement Active Oberon 2019? Active Oberon 2019 is **_another language_**. Most if not all Oberon compilers on Github only implement **_original Oberon_**. I don't see Active Oberon 2019 has anything superior than what Oberon+ already has.

**Reply by nsgnkhibdk2cls0f** [reply]
IMO Oberon is just a brand now. They do anything to sell it to the public, an Oberon-2 even called itself Component Pascal, of course there is no Pascal in it, Pascal is only used as a brand. Same as Oberon now, just a brand, a marketing buzzword. Original Oberon's latest version is Oberon-07. Active Oberon, Zonnon,... are other derived languages that used the brand Oberon.

**Reply by nsgnkhibdk2cls0f** [reply]
Active Oberon language report: https://gitlab.inf.ethz.ch/felixf/oberon/-/blob/main/docu/OberonLanguageReport.pdf

I don't see any breakthroughs. The life and evolution of Oberon is just you trimmed the language too much so you have to re-add it, then you feel it's **_bloated_** so you started to trim it down even further only to realize your language is now **_so useless_** and you have to re-add something you removed back and this keeps repeating. Because Original Oberon is **_too useless_** as it was trimmed down too much, derivatives like Active Oberon, Zonnon (derived from Active Oberon), Component Pascal (derived from Oberon-2) have to add some of the features back and they called it as improvements!

**Reply by 1997tn** [reply]
@dmuysers If you liked Oberon, there are a lot of Oberons on oberon.org for you to choose from :)
